pkg/agent/attestor/workload/workload.go

```diff
@@ -0,0 +1,91 @@
+package attestor
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/pkg/agent/catalog"
+	"github.com/spiffe/spire/pkg/common/telemetry"
+	"github.com/spiffe/spire/proto/agent/workloadattestor"
+	"github.com/spiffe/spire/proto/common"
+)
+
+type attestor struct {
+	c *Config
+}
+
+type Attestor interface {
+	Attest(pid int32) []*common.Selector
+}
+
+func New(config *Config) Attestor {
+	return &attestor{c: config}
+}
+
+type Config struct {
+	Catalog catalog.Catalog
+	L       logrus.FieldLogger
+	T       telemetry.Sink
+}
+
+// Attest invokes all workload attestor plugins against the provided PID. If an error
+// is encountered, it is logged and selectors from the failing plugin are discarded.
+func (wla *attestor) Attest(pid int32) []*common.Selector {
+	tLabels := []telemetry.Label{{"workload_pid", string(pid)}}
+	defer wla.c.T.MeasureSinceWithLabels([]string{"workload_api", "workload_attestation_duration"}, time.Now(), tLabels)
+
+	plugins := wla.c.Catalog.WorkloadAttestors()
+	sChan := make(chan []*common.Selector)
+	errChan := make(chan error)
+
+	for _, p := range plugins {
+		go wla.invokeAttestor(p, pid, sChan, errChan)
+	}
+
+	// Collect the results
+	selectors := []*common.Selector{}
+	for i := 0; i < len(plugins); i++ {
+		select {
+		case s := <-sChan:
+			selectors = append(selectors, s...)
+		case err := <-errChan:
+			wla.c.L.Errorf("Failed to collect all selectors for PID %v: %v", pid, err)
+		}
+	}
+
+	wla.c.T.AddSampleWithLabels([]string{"workload_api", "discovered_selectors"}, float32(len(selectors)), tLabels)
+	wla.c.L.Debugf("PID %v attested to have selectors %v", pid, selectors)
+	return selectors
+}
+
+// invokeAttestor invokes attestation against the supplied plugin. Should be called from a goroutine.
+func (wla *attestor) invokeAttestor(a workloadattestor.WorkloadAttestor, pid int32, sChan chan []*common.Selector, errChan chan error) {
+	attestorName := wla.attestorName(a)
+	tLabels := []telemetry.Label{{"workload_pid", string(pid)}, {"attestor_name", attestorName}}
+
+	req := &workloadattestor.AttestRequest{
+		Pid: pid,
+	}
+
+	start := time.Now()
+	resp, err := a.Attest(req)
+	wla.c.T.MeasureSinceWithLabels([]string{"workload_api", "workload_attestor_latency"}, start, tLabels)
+	if err != nil {
```

Comment:
must not we validate error just after Attest? or we need something from "MeasureSinceWithLabels"?

Reply:
Most of this code is copy/pasted from the existing workload attestor logic. I put the error check after this line because I thought it would be beneficial to capture the attestor latency metrics regardless of whether an error condition was encountered or not